extensions/classify/config.proto

```diff
@@ -0,0 +1,198 @@
+message Value {
+  // Assign `value` to the enclosing dimension if match is true.
+  string value = 1;
+
+  // Match specification.
+  // Empty match always matches, which can be used to specify a default value when placed
+  // at the end of the list.
+  Match match = 2;
+}
+
+// Match condition based on request and response information.
+// Every match condition must match for the result to be true.
+message Match {
+  // invert the match condition.
+  bool invert = 1;
+
+  // Match path.
+  repeated StringMatch path = 2;
+
+  // Match scheme: http, https or unset.
+  Scheme scheme = 3;
+
+  // Match method: GET, POST, HEAD, PUT, DELETE, ...
+  repeated string method = 4;
+
+  // Match request headers.
+  // Not implemented.
+  // $hide_from_docs
+  map<string, StringMatch> request_headers = 5;
+  
+  // Match response headers.
+  // Not implemented.
+  // $hide_from_docs
+  map<string, StringMatch> response_headers = 6;
+
+  // Match query parameters.
+  // Not implemented.
+  // $hide_from_docs
+  map<string, StringMatch> query_params = 7;
+
+  // Match response code.
+  repeated IntMatch response_code = 8;
+
+  // Match request size.
+  // Not implemented.
+  // $hide_from_docs
+  repeated IntMatch request_size = 9;
+
+  // Match response size.
+  // Not implemented.
+  // $hide_from_docs
+  repeated IntMatch response_size = 10;
+
+  // Match response duration.
+  // Not implemented.
+  // $hide_from_docs
+  repeated DurationMatch response_duration = 11;
+}
+
+// Describes how to match a given string.
+message StringMatch {
+  // default is case sensitive.
+  bool case_insensitive = 1;
+
+  oneof match_type {
+
+    // exact string match
+    string exact = 2;
+
+    // prefix-based match
+    string prefix = 3;
+
+    // RE2 style regex-based match (https://github.com/google/re2/wiki/Syntax).
+    string regex = 4;
+  }
+}
+
+// Describes how to match a given Integer.
+message IntMatch {
+  oneof match_type {
+
+    // exact Int match
+    int64 exact = 1;
+
+    // Range match
+    IntRange range = 2;
+  }
+}
+
+// Describes how to match a Duration.
+message DurationMatch {
+  oneof match_type {
+
+    // exact duration match.
+    google.protobuf.Duration exact = 1;
+
+    // Range match for duration.
+    DurationRange range = 2;
+  }
+}
+
+// IntRange matches a range of integers.
+message IntRange {
+  // Greater than or equal to this number.
+  // Default is 0.
+  int64 min = 1;
+
+  // Less than or equal to this number.
+  // Default is INT_MAX.
+  int64 max = 2;
+}
+
+// DurationRange matches a range of integers.
+message DurationRange {
+  // Greater than or equal to this duration.
+  // Default is 0.
+  google.protobuf.Duration min = 1;
+
+  // Less than or equal to this duration.
+  // Default is INT_MAX.
+  google.protobuf.Duration max = 2;
+}
+
+enum Scheme {
+  // Matches any
+  UNSPECIFIED = 0;
+  HTTP = 1;
+  HTTPS = 2;
+}
+
+// Phase determines when the classification is performed.
```

Comment:
Does Phase capture intent of the user or actually places the filter at the right place i.e. do users still need to create EnvoyFilter with the current placement settings for this new filter?

Reply:
User needs to place the filter correctly in relation to other filters.
Phase dictates the filter callback when classification is performed.

Reply:
Isn't that too low level for any user?

Why can't we have the Phase fixed for the most common use cases as most of the end users won't understand this.

Reply:
You cannot have a fixed phase. If I want to use the output of classification for responseClass in the stats filter then the classification must be performed in LOG phase when responseCodes are available.

Alternatively if you want to use the classification in an authz filter you cannot perform classification in the LOG phase, it will be too late and authz will not have the request classification that is needed.

We can do some magic to figure this out, I think we should start simple and explicit .

Reply:
Can the filter guess based on its position? It's not a good experience to both specify the filter position in EnvoyFilter, and then specify it again in the phase of the filter config.